Sources/SwiftProtobuf/JSONScanner.swift

```diff
@@ -706,6 +398,311 @@ internal struct JSONScanner {
+          throw JSONDecodingError.leadingZero
+        case asciiPeriod, asciiLowerE, asciiUpperE: // . e
+          // Slow path: JSON numbers can be written in floating-point notation
+          index = start
+          if let d = try parseBareDouble(source: source,
+                                         index: &index,
+                                         end: end) {
+            if let u = UInt64(exactly: d) {
+              return u
+            }
+          }
+          throw JSONDecodingError.malformedNumber
+        case asciiBackslash:
+          return nil
+        default:
+          return 0
+        }
+      }
+      return 0
+    case asciiOne...asciiNine: // 1...9
+      var n = UInt64(c - 48)
+      while index != end {
+        let digit = source[index]
+        switch digit {
+        case asciiZero...asciiNine: // 0...9
+          let val = UInt64(digit - asciiZero)
+          if n >= UInt64.max / 10 {
+            if n > UInt64.max / 10 || val > UInt64.max % 10 {
+              throw JSONDecodingError.numberRange
+            }
+          }
+          source.formIndex(after: &index)
+          n = n * 10 + val
+        case asciiPeriod, asciiLowerE, asciiUpperE: // . e
+          // Slow path: JSON allows floating-point notation for integers
+          index = start
+          if let d = try parseBareDouble(source: source,
+                                         index: &index,
+                                         end: end) {
+            if let u = UInt64(exactly: d) {
+              return u
+            }
+          }
+          throw JSONDecodingError.malformedNumber
+        case asciiBackslash:
+          return nil
+        default:
+          return n
+        }
+      }
+      return n
+    case asciiBackslash:
+      return nil
+    default:
+      throw JSONDecodingError.malformedNumber
+    }
+  }
+
+  // Parse the leading Int64 from the provided utf8.
+  //
+  // This uses parseBareUInt() to do the heavy lifting;
+  // we just check for a leading minus and negate the result
+  // as necessary.
+  //
+  // As with parseBareUInt(), if it encounters a "\" backslash
+  // character, it returns a nil.  This is used by callers that are
+  // parsing quoted numbers.  See nextSInt() and nextUInt() below.
+
+  private func parseBareSInt(
+    source: UnsafeBufferPointer<UInt8>,
+    index: inout UnsafeBufferPointer<UInt8>.Index,
+    end: UnsafeBufferPointer<UInt8>.Index
+  ) throws -> Int64? {
+    if index == end {
+      throw JSONDecodingError.truncated
+    }
+    let c = source[index]
+    if c == asciiMinus { // -
+      source.formIndex(after: &index)
+      // character after '-' must be digit
+      let digit = source[index]
+      if digit < asciiZero || digit > asciiNine {
+        throw JSONDecodingError.malformedNumber
+      }
+      if let n = try parseBareUInt(source: source, index: &index, end: end) {
+        let limit: UInt64 = 0x8000000000000000 // -Int64.min
```

Comment:
I could write this in other ways, but I'm not sure what you would find clearer.  Which of these would you prefer?
```
   UInt64(1) << UInt64(63)
   UInt64(Int64.max) + 1
```